src/options.rs

```diff
@@ -0,0 +1,138 @@
+use crate::config::ConfigFile;
+use crate::error::Error;
+use crate::repo::{BlockStore, DataStore};
+use async_std::path::PathBuf;
+use bitswap::Strategy;
+use libp2p::core::{Multiaddr, PeerId};
+use libp2p::identity::Keypair;
+use std::borrow::Borrow;
+use std::fmt;
+
+/// All types can be changed at compile time by implementing
+/// `IpfsTypes`.
+pub trait IpfsTypes: Clone + 'static {
+    type TBlockStore: BlockStore;
+    type TDataStore: DataStore;
+    type TStrategy: Strategy;
+}
+
+/// Default IPFS types.
+#[derive(Clone, Debug)]
+pub struct Types;
+impl IpfsTypes for Types {
+    type TBlockStore = crate::repo::fs::FsBlockStore;
+    #[cfg(feature = "rocksdb")]
+    type TDataStore = crate::repo::fs::RocksDataStore;
+    #[cfg(not(feature = "rocksdb"))]
+    type TDataStore = crate::repo::mem::MemDataStore;
+
+    type TStrategy = bitswap::AltruisticStrategy;
+}
+
+/// Testing IPFS types
+#[derive(Clone, Debug)]
+pub struct TestTypes;
+impl IpfsTypes for TestTypes {
+    type TBlockStore = crate::repo::mem::MemBlockStore;
+    type TDataStore = crate::repo::mem::MemDataStore;
+    type TStrategy = bitswap::AltruisticStrategy;
+}
+
+/// Ipfs options
+#[derive(Clone)]
+pub struct IpfsOptions {
+    /// The path of the ipfs repo.
+    pub ipfs_path: PathBuf,
+    /// The keypair used with libp2p.
+    pub keypair: Keypair,
+    /// Nodes dialed during startup.
+    pub bootstrap: Vec<(Multiaddr, PeerId)>,
+    /// Enables mdns for peer discovery when true.
+    pub mdns: bool,
+}
+
+impl fmt::Debug for IpfsOptions {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // needed since libp2p::identity::Keypair does not have a Debug impl, and the IpfsOptions
+        // is a struct with all public fields, so don't enforce users to use this wrapper.
+        fmt.debug_struct("IpfsOptions")
+            .field("ipfs_path", &self.ipfs_path)
+            .field("bootstrap", &self.bootstrap)
+            .field("keypair", &DebuggableKeypair(&self.keypair))
+            .field("mdns", &self.mdns)
+            .finish()
+    }
+}
+
+/// Workaround for libp2p::identity::Keypair missing a Debug impl, works with references and owned
+/// keypairs.
+#[derive(Clone)]
+pub(crate) struct DebuggableKeypair<I: Borrow<Keypair>>(pub I);
+
+impl<I: Borrow<Keypair>> fmt::Debug for DebuggableKeypair<I> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let kind = match self.as_ref() {
+            Keypair::Ed25519(_) => "Ed25519",
+            Keypair::Rsa(_) => "Rsa",
+            Keypair::Secp256k1(_) => "Secp256k1",
+        };
+
+        write!(fmt, "Keypair::{}", kind)
+    }
+}
+
+impl<I: Borrow<Keypair>> AsRef<Keypair> for DebuggableKeypair<I> {
+    fn as_ref(&self) -> &Keypair {
+        self.0.borrow()
+    }
+}
```

Comment:
I am glad you found a way for this to work, I feel I must've tried out this configuration as well but always ran into "needs type hint" issue!